rectools/models/nn/transformer_base.py

```diff
@@ -0,0 +1,517 @@
+            gbce_t=self.gbce_t,
+            n_item_extra_tokens=n_item_extra_tokens,
+        )
+
+        self.fit_trainer = deepcopy(self._trainer)
+        self.fit_trainer.fit(self.lightning_model, train_dataloader)
+
+    def _custom_transform_dataset_u2i(
+        self, dataset: Dataset, users: ExternalIds, on_unsupported_targets: ErrorBehaviour
+    ) -> Dataset:
+        return self.data_preparator.transform_dataset_u2i(dataset, users)
+
+    def _custom_transform_dataset_i2i(
+        self, dataset: Dataset, target_items: ExternalIds, on_unsupported_targets: ErrorBehaviour
+    ) -> Dataset:
+        return self.data_preparator.transform_dataset_i2i(dataset)
+
+    def _recommend_u2i(
+        self,
+        user_ids: InternalIdsArray,
+        dataset: Dataset,  # [n_rec_users x n_items + n_item_extra_tokens]
+        k: int,
+        filter_viewed: bool,
+        sorted_item_ids_to_recommend: tp.Optional[InternalIdsArray],  # model_internal
+    ) -> InternalRecoTriplet:
+        if sorted_item_ids_to_recommend is None:  # TODO: move to _get_sorted_item_ids_to_recommend
+            sorted_item_ids_to_recommend = self.data_preparator.get_known_items_sorted_internal_ids()  # model internal
+
+        recommend_trainer = Trainer(devices=1, accelerator=self.recommend_device)
+        recommend_dataloader = self.data_preparator.get_dataloader_recommend(dataset)
+        session_embs = recommend_trainer.predict(model=self.lightning_model, dataloaders=recommend_dataloader)
+        if session_embs is not None:
+            user_embs = np.concatenate(session_embs, axis=0)
+            user_embs = user_embs[user_ids]
+            item_embs = self.lightning_model.item_embs
+            item_embs_np = item_embs.detach().cpu().numpy()
+
+            ranker = ImplicitRanker(
+                self.u2i_dist,
+                user_embs,  # [n_rec_users, n_factors]
+                item_embs_np,  # [n_items + n_item_extra_tokens, n_factors]
+            )
+            if filter_viewed:
+                user_items = dataset.get_user_item_matrix(include_weights=False)
+                ui_csr_for_filter = user_items[user_ids]
+            else:
+                ui_csr_for_filter = None
+
+            # TODO: When filter_viewed is not needed and user has GPU, torch DOT and topk should be faster
+            user_ids_indices, all_reco_ids, all_scores = ranker.rank(
+                subject_ids=np.arange(user_embs.shape[0]),  # n_rec_users
+                k=k,
+                filter_pairs_csr=ui_csr_for_filter,  # [n_rec_users x n_items + n_item_extra_tokens]
+                sorted_object_whitelist=sorted_item_ids_to_recommend,  # model_internal
+                num_threads=self.recommend_n_threads,
+                use_gpu=self.recommend_use_gpu_ranking and HAS_CUDA,
+            )
+            all_target_ids = user_ids[user_ids_indices]
+        else:
+            explanation = """Received empty recommendations. Used for type-annotation"""
```

Comment:
1. I don't really understand the second part
2. Let's swap if and else cases (we don't need else case then)